blinemedical/amazon-s3-gst-plugin
# [VIDEO-2016] extend 'seek' feature set

**VIDEO-2016**

The specific use case / issues being solved were that when writing video files, we needed to seek back to the head of the file to write a few bytes of data and then finish.  This involved several CURL calls to the AWS API:

1. **Upload what we have**
2. **Complete (destroy the uploader / finalize the file)**
3. **Download from 0->offset**
4. write some data
5. **Download from (2)+(3)+1 up to the end of a part minimum size (i.e., 5 MB)**
6. **Upload that**
7. **Either Download->Upload** the rest of the file or **Copy-upload** the rest of the file (2 or 1 CURL respectively)
8. **Complete (destroy the uploader / finalize the file)**

Step 5 represents the entry point for end of stream.  Each of these API calls (excluding step 4 which isn't an API call) would take a fraction of a second up to several seconds each, quickly consuming our available response window of 9.9 seconds (so, 4 or 5 calls).

This change set takes that above scenario down to the following:
1. **Upload what we have**
2. "soft seek" to the cache (no API call)
3. write some data
4. **Upload the affected part again**
5. <detect the uploader already knows of the end of file and "soft seek" to the EOF>
6. **Complete** (destroy the uploader / finalize the file

Step 4 represents the entry point for end of stream.  This change set results in an EOS API call reduction from 4-5 down to 2 with only one of those steps actually moving data.

* Updated existing tests function
* Added a "part cache" feature that can track some number of stored buffers within a greater set of information (e.g., the other part sizes).  Storage and clean-up of the buffers happens on insertion and either tracks N buffers at the head or tail of the total list.
* Extended GstS3Uploader API to include a 'seek' method that takes an offset and (potentially) returns a copy of the buffer and its size from the cache, if found, or just the size if known.
* Extended the s3sink to include `num-cache-parts`, supporting [-10000,0,10000] as a means to utilize the cache from gstreamer.
* Added the `s3sink-cached` test suite to perform high-fidelity tests on new test cases surrounding the use of the cache to increase confidence in future modifications.

**TEST PROCEDURE**

Compile and run `make test [-C builddir]`.  Everything should pass.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
